src/SubscriptionBase.h

```diff
@@ -0,0 +1,31 @@
+// Copyright 2004-present Facebook. All Rights Reserved.
+
+#pragma once
+
+#include <folly/MoveWrapper.h>
+#include "src/EnableSharedFromThis.h"
+#include "src/ReactiveStreamsCompat.h"
+#include "src/mixins/ExecutorMixin.h"
+
+namespace reactivesocket {
+
+class SubscriptionBase : public Subscription,
+                         public EnableSharedFromThisBase<SubscriptionBase>,
+                         public virtual ExecutorBase {
+  virtual void requestImpl(size_t n) = 0;
+  virtual void cancelImpl() = 0;
+
+ public:
+  using ExecutorBase::ExecutorBase;
+
+  void request(size_t n) override final {
+    runInExecutor(
+        std::bind(&SubscriptionBase::requestImpl, shared_from_this(), n));
```

Comment:
In the constructor the internal weak_ptr is not set yet. It is set by the shared_ptr afterwards.
So the behavior that you will always get is exception bad_weak_ptr. Unfortunately that's what we have to live with ;(